feat(parser): Add parser selection with json parser

MrKevinWeiss opened the pull request:
This introduces a parser selection to help with versioning.  It also comes with a base shell if the DUT uses the helper functions.  Also some general cleanup is done here...

### Testing Procedure
- Flash a board with the  [app_metadata test](https://github.com/RIOT-OS/RobotFW-tests/pull/28) from the RobotFW repo
- Update riot_pal to 0.3.0 locally
- Run `riot_pal/dut_pyshell.py`
- Try each option that can be autocompleted from `send_cmd <command>`
- Note that some arguments can be added such as ints or strings

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+            print('Select a serial port:')
+            for i, s_dev in enumerate(serial_devices):
+                print("{}: {}".format(i, s_dev))
+            s_num = input("Selection(number): ")
```
this and the next line are not very safe, I think something like this might be better:

```Python
try:
    s_num = int(input("Selection(number): "))
except:
    #error message?
    raise ValueError
```

MrKevinWeiss replied:
Agreed

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+    Args:
+        port - Serial port for the PHiLIP, if None connection wizard tries to
+               connent
+        data_only - If true only data prints from command an not the whole
```
why not call this `debug` and invert the logic? At least to me that is what other programs do.

MrKevinWeiss replied:
Heh it should actually be called filter data.
I wanted debug to be a bit more verbose and unstructured.  Without the filter it is very structured and with the filter it is more human readable.

smlng replied:
mhm, okay maybe I need to look at it again and rethink ...

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+                    help='Python logger log level')
+PARSER.add_argument('--port', '-p', help='Specifies the serial port',
+                    default=None)
+PARSER.add_argument('--data_filter_off', '-do', default=True,
```
all short hands short be single letters, and as mentioned above a better name for this would be `debug` hence `-d`

MrKevinWeiss replied:
I agree names could improve

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+PARSER.add_argument('--data_filter_off', '-do', default=True,
+                    action='store_false',
+                    help='Filters data from philip responses')
+PARSER.add_argument('--use_dev_map', '-dm', default=False,
```
why not `memmap` and `-m`

MrKevinWeiss replied:
heh the real question is what is this doing in there.  I need to cleanup this alot (I think only philip can self describe its map).

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+PARSER = argparse.ArgumentParser()
+
+LOG_LEVELS = ('debug', 'info', 'warning', 'error', 'fatal', 'critical')
+PARSER.add_argument('--loglevel', choices=LOG_LEVELS, default='info',
```
could add `-l` (lower L) as short hand

MrKevinWeiss replied:
sexy.  Maybe I will also tell @cladmi to do that in his scripts :wink:

cladmi replied:
Tools usually refers to `-d` for the short option for debug level or to `-v` as reference to verbosity more than `-l`.

smlng commented on riot_pal/dut_shell.py:
```diff
@@ -57,18 +66,82 @@ def send_cmd(self, send_cmd):
+        return cmd_info
+
+
+class DutShell:
```
run this through flake8 or a linter, it will complain about missing doc strings 😉

MrKevinWeiss replied:
The most important, documentation!

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+
+    def _print_func_result(self, func, arg):
+        value = (arg or '').split(' ')
+        func_args = [v for v in value if v]
```
can be omitted bc `value == func_args`, so simply use `value` (or better name it `values` or `func_args` directly.

MrKevinWeiss replied:
Does this filter out all empty values (if you have two spaces)? It was copy pasta... I will change the names and test.

smlng replied:
ah you're right, currently it does not but it works if you use
```Python
value = arg.split()
```

bc `split` without specifying the separator split by any number of spaces, and does not create empty items.

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+        for res in results:
+            if self.data_only:
+                if 'result' in res:
+                    if res['result'] == RESULT_SUCCESS:
```
why this distinction, if there was an error data should be empty and printed accordingly.

MrKevinWeiss replied:
I am not super sure about that.  How does one get error codes?  It is all about filtering only the minimum and most useful data.  One can turn off the filter if they need to know the error code.

smlng commented on riot_pal/dut_pyshell.py:
```diff
@@ -0,0 +1,258 @@
+
+    def complete_send_cmd(self, text, line, begidx, endidx):
+        """Completes arg with memory map record names"""
+        begidx = begidx
```
is this just a dummy function?

MrKevinWeiss replied:
Nope, this allows autocomplete of functions, at least if the help is structured!

smlng replied:
okay

smlng reviewed (changes requested):
first review round

MrKevinWeiss commented:
updated!

smlng reviewed (approved):
please squash

MrKevinWeiss committed:
chore(tests/helpers): Fix pylint errors

MrKevinWeiss committed:
feat(parser): Add parser selection with json parser

Add the option to select a parser
Add json parser implementation
This should help with backwards compatibility
Shell is the old style which was just a convention

MrKevinWeiss committed:
feat(dut_pyshell): Add a python shell wrapper for dev tests

This should help developers test if using the json parser
It allows autocomplete and command list discovery

MrKevinWeiss committed:
chore(setup): Bump version and add dut_pyshell console script

MrKevinWeiss commented:
Done

MrKevinWeiss commented:
Is this ready to go.  When merged I will then upload to pip.

MrKevinWeiss commented:
Pip version 0.3.0 released.
